### PR TITLE
Define variable for minimun replication factor

### PR DESCRIPTION
**Pull request by jeqo**
# Description

Currently, replication factor to be used for internal topics is calculated by `min(#brokers, 3)`. 
Even though this works in most cases, when deploying Kafka cluster in 2 DC another minimun can be required: e.g. running 8 brokers, 4/DC, can require RF=4 and min.isr=2. 

Changes:

- Adding a default internal replication factor.
- Add RF for Schema Registry topic.
- Add RF for Connect Secret Registry. 

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Updating one test-case to check this scenario.

# Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] Any variable changes have been validated to be backwards compatible

### CODE REVIEW

File: roles/confluent.variables/vars/main.yml [hunk 1]
```diff
@@ -394,6 +394,7 @@ schema_registry_properties:
       debug: 'false'
       schema.registry.group.id: schema-registry
       kafkastore.topic: _schemas
+      kafkastore.topic.replication.factor: "{{ [ groups['kafka_broker'] | default(['localhost']) | length, default_internal_replication_factor ] | min }}"
```
Comment (domenicbove) [inline roles/confluent.variables/vars/main.yml:397]: should we create a `schema_registry_default_internal_replication_factor` var then, seems like thats the pattern w the other components?

Comment (jeqo) [reply roles/confluent.variables/vars/main.yml:397]: Yeap, make sense. Will add it

File: roles/confluent.test/molecule/mtls-debian/molecule.yml [hunk 2]
```diff
@@ -122,6 +134,8 @@ provisioner:
   inventory:
     group_vars:
       all:
+        default_internal_replication_factor: 4
```
Comment (domenicbove) [inline roles/confluent.test/molecule/mtls-debian/molecule.yml:137]: So the value of this PR is to make it easier to set the RF across all components? Bc you could just use the existing variables right?
```
kafka_broker_default_internal_replication_factor: 4
ksql_default_internal_replication_factor: 4
control_center_default_internal_replication_factor: 4
```

Comment (jeqo) [reply roles/confluent.test/molecule/mtls-debian/molecule.yml:137]: Good point. At the beginning, I didn't know how many variables to change to align the whole services.

We can seize if it's worth adding this as a common variable.

Comparing it to other scenarios as license being set per component, we can delegate this global variable to the user and update the RF per service at the playbook level. These can be added as a use-case to the docs.

If you share this, I can rollback changes and reduce the scope to only adding the SR property, and look into the docs to get this use-case cover.

Comment (domenicbove) [reply roles/confluent.test/molecule/mtls-debian/molecule.yml:137]: No I'm cool with this bc the 3 should not be hardcoded everywhere anyhow :)

And UX wise, this: 
```
default_internal_replication_factor: 4
```

is easier than 
```
kafka_broker_default_internal_replication_factor: 4
kafka_connect_default_internal_replication_factor: 4
ksql_default_internal_replication_factor: 4
control_center_default_internal_replication_factor: 4
```

This is similar to making a global variable for the confluent.license property, we have that on our roadmap

Comment (jeqo) [reply roles/confluent.test/molecule/mtls-debian/molecule.yml:137]: Awesome! Thanks for the tip also, will use it with the current version.

Comment (domenicbove) [inline roles/confluent.test/molecule/mtls-debian/molecule.yml:137]: Oh one more thing, can you add some tasks to the verify.yml of this scenario that validates the variable does what you expect? We have a helper task file called check property.

You can run the verify.yml playbook after you converge with: 
```
molecule verify -s mtls-debian
```

**Comment by OoRobFuoO** [comment]
@jeqo I would suggest adding a condition to apply this only if there're > 4 brokers in the cluster.
 There're many people who just have a 3 Kafka broker cluster, so this change won't work for them.

**Comment by domenicbove** [comment]
@OoRobFuoO - the default remains 3, you have to customize the var to set it to 4

File: roles/confluent.test/molecule/mtls-debian/molecule.yml (hunk 2 above)
Comment (jeqo) [reply roles/confluent.test/molecule/mtls-debian/molecule.yml:137]: added.

ps: as now jolokia_enabled=false by default, molecule tests that expect jolokia available will fail. will create a issue to track this.

Comment (jeqo) [reply roles/confluent.test/molecule/mtls-debian/molecule.yml:137]: https://github.com/confluentinc/cp-ansible/issues/532

File: roles/confluent.test/molecule/mtls-debian/verify.yml [hunk 3]
```diff
@@ -10,6 +10,13 @@
         file_path: /etc/kafka/server.properties
         property: listener.security.protocol.map
         expected_value: INTERNAL:SSL,BROKER:SSL
+    - import_role:
```
Comment (domenicbove) [inline roles/confluent.test/molecule/mtls-debian/verify.yml:13]: Can we get some line breaks in this file?

**Review by domenicbove — APPROVED** [review]
Looks great, one final comment for line breaks!

File: VARIABLES.md [hunk 4]
```diff
@@ -220,6 +220,14 @@ Default:  /usr/local/bin/confluent
 
 ***
 
+### default_internal_replication_factor
+
+Recommended replication factor. Usually 3. If running in 2DC and look for balanced replicad, consider increasing to 4. If number of brokers is less than this property, then minimun value is used.
```
Comment (JumaX) [inline VARIABLES.md:225]:
```suggestion
Recommended replication factor, defaults to 3. When splitting your cluster across 2 DCs with 4 or more Brokers, this should be increased to 4 to balance topic replicas.```

File: VARIABLES.md [hunk 5]
```diff
@@ -654,9 +662,9 @@ Default:  []
 
 ### kafka_broker_default_internal_replication_factor
 
-Replication Factor for internal topics. Defaults to the minimum of the number of brokers and 3
+Replication Factor for internal topics. Defaults to the minimum of the number of brokers and default replication factor (see default_internal_replication_factor)
```
Comment (JumaX) [inline VARIABLES.md:665]:
```suggestion
Replication Factor for internal topics. Defaults to the minimum of the number of brokers and can be overridden via  default replication factor (see default_internal_replication_factor)
```

File: VARIABLES.md [hunk 6]
```diff
@@ -716,6 +724,14 @@ Default:  8081
 
 ***
 
+### schema_registry_default_internal_replication_factor
+
+Replication Factor for schemas topic. Defaults to the minimum of the number of brokers and default replication factor (see default_internal_replication_factor)
```
Comment (JumaX) [inline VARIABLES.md:729]:
```suggestion
Replication Factor for schemas topic. Defaults to the minimum of the number of brokers and can be overridden via  default replication factor (see default_internal_replication_factor)
```

File: VARIABLES.md [hunk 7]
```diff
@@ -1118,9 +1134,9 @@ Default:  connect-cluster
 
 ### kafka_connect_default_internal_replication_factor
 
-Replication Factor for connect internal topics. Defaults to the minimum of the number of brokers and 3
+Replication Factor for connect internal topics. Defaults to the minimum of the number of brokers and default replication factor (see default_internal_replication_factor)
```
Comment (JumaX) [inline VARIABLES.md:1137]:
```suggestion
Replication Factor for connect internal topics. Defaults to the minimum of the number of brokers and can be overridden via  default replication factor (see default_internal_replication_factor).
```

File: VARIABLES.md [hunk 8]
```diff
@@ -1294,9 +1310,9 @@ Default:  []
 
 ### ksql_default_internal_replication_factor
 
-Replication Factor for ksqlDB internal topics. Defaults to the minimum of the number of brokers and 3
+Replication Factor for ksqlDB internal topics. Defaults to the minimum of the number of brokers and default replication factor (see default_internal_replication_factor)
```
Comment (JumaX) [inline VARIABLES.md:1313]:
```suggestion
Replication Factor for ksqlDB internal topics. Defaults to the minimum of the number of brokers and can be overridden via  default replication factor (see default_internal_replication_factor).
```

File: VARIABLES.md [hunk 9]
```diff
@@ -1398,9 +1414,9 @@ Default:  []
 
 ### control_center_default_internal_replication_factor
 
-Replication Factor for Control Center internal topics. Defaults to the minimum of the number of brokers and 3
+Replication Factor for Control Center internal topics. Defaults to the minimum of the number of brokers and default replication factor (see default_internal_replication_factor)
```
Comment (JumaX) [inline VARIABLES.md:1417]:
```suggestion
Replication Factor for Control Center internal topics. Defaults to the minimum of the number of brokers and can be overridden via  default replication factor (see default_internal_replication_factor).
```

File: roles/confluent.variables/defaults/main.yml [hunk 10]
```diff
@@ -107,6 +107,9 @@ confluent_cli_download_enabled: "{{rbac_enabled or secrets_protection_enabled}}"
 ### Full path on hosts to install the Confluent CLI
 confluent_cli_path: /usr/local/bin/confluent
 
+### Recommended replication factor. Usually 3. If running in 2DC and look for balanced replicad, consider increasing to 4. If number of brokers is less than this property, then minimun value is used.
```
Comment (JumaX) [inline roles/confluent.variables/defaults/main.yml:110]:
```suggestion
### Recommended replication factor, defaults to 3. When splitting your cluster across 2 DCs with 4 or more Brokers, this should be increased to 4 to balance topic replicas.```
```

File: roles/confluent.variables/defaults/main.yml [hunk 11]
```diff
@@ -608,8 +614,8 @@ kafka_connect_copy_files: []
 ### Connect Service Group Id. Customize when configuring multiple connect clusters in same inventory
 kafka_connect_group_id: connect-cluster
 
-### Replication Factor for connect internal topics. Defaults to the minimum of the number of brokers and 3
-kafka_connect_default_internal_replication_factor: "{{ [ groups['kafka_broker'] | default(['localhost']) | length, 3 ] | min }}"
+### Replication Factor for connect internal topics. Defaults to the minimum of the number of brokers and default replication factor (see default_internal_replication_factor)
```
Comment (JumaX) [inline roles/confluent.variables/defaults/main.yml:617]: See previous comment suggestions on how to modify this.

File: roles/confluent.variables/defaults/main.yml [hunk 12]
```diff
@@ -703,8 +709,8 @@ ksql_jmxexporter_port: 8076
 ### Use to copy files from control node to ksqlDB hosts. Set to list of dictionaries with keys: source_path (full path of file on control node) and destination_path (full path to copy file to)
 ksql_copy_files: []
 
-### Replication Factor for ksqlDB internal topics. Defaults to the minimum of the number of brokers and 3
-ksql_default_internal_replication_factor: "{{ [ groups['kafka_broker'] | default(['localhost']) | length, 3 ] | min }}"
+### Replication Factor for ksqlDB internal topics. Defaults to the minimum of the number of brokers and default replication factor (see default_internal_replication_factor)
```
Comment (JumaX) [inline roles/confluent.variables/defaults/main.yml:712]: See previous comment suggestions on how to modify this.

File: roles/confluent.variables/defaults/main.yml [hunk 13]
```diff
@@ -763,8 +769,8 @@ control_center:
 ### Use to copy files from control node to Control Center hosts. Set to list of dictionaries with keys: source_path (full path of file on control node) and destination_path (full path to copy file to)
 control_center_copy_files: []
 
-### Replication Factor for Control Center internal topics. Defaults to the minimum of the number of brokers and 3
-control_center_default_internal_replication_factor: "{{ [ groups['kafka_broker'] | default(['localhost']) | length, 3 ] | min }}"
+### Replication Factor for Control Center internal topics. Defaults to the minimum of the number of brokers and default replication factor (see default_internal_replication_factor)
```
Comment (JumaX) [inline roles/confluent.variables/defaults/main.yml:772]: See previous comment suggestions on how to modify this.

**Review by JumaX — CHANGES_REQUESTED** [review]
Some spelling and grammar issues in the comments.

File: VARIABLES.md (hunk 4 above)
Comment (domenicbove) [reply VARIABLES.md:225]: Make sure these changes get reflected in both the VARIABLES.md and the place annotation originates

**Comment by jeqo** [comment]
UPDATE: Added RF for Connect Secret Registry as well.

**Review by JumaX — APPROVED** [review]
LGTM.

**Comment by JumaX** [comment]
@jeqo Looks like there maybe a linting errors, you will need to resolve those before the PR can be submitted.

**Comment by jeqo** [comment]
@JumaX I can't reproduce any linting issue locally. 
https://jenkins.confluent.io/job/confluentinc-pr/job/cp-ansible/job/PR-531/6/flowGraphTable/ doesn't show any exception from the source code neither. 
Could we re-run the pipeline?

**Comment by JumaX** [comment]
@jeqo There is an issue with Jenkins at the moment, I am unable to re-run the test. I've reached out for additional support and will update you once resolved.

**Comment by domenicbove** [comment]
@jeqo does the test pass when you run it locally?

**Comment by JumaX** [comment]
@jeqo it looks like there are linting issues still:

```
18:21:07  yamllint -c ../../.yamllint ../..[0m
18:21:12  ../../all.yml
18:21:12    38:1      error    syntax error: could not find expected ':' (syntax)
18:21:12  
18:21:12  ../../roles/confluent.test/molecule/plain-rhel/verify.yml
18:21:12    104:18    error    syntax error: could not find expected ':' (syntax)
```

**Comment by jeqo** [comment]
@JumaX I haven't touch the files you list with issues:
```
~/src/github.com/confluentinc/cp-ansible min-rf ········································································ 19:53:16
❯ git status
On branch min-rf
nothing to commit, working tree clean

~/src/github.com/confluentinc/cp-ansible min-rf ········································································ 19:53:24
❯ yamllint -c .yamllint .

~/src/github.com/confluentinc/cp-ansible min-rf ····································································· 7s 19:53:33
❯ 
```

**Pull request merged**